apps/mark-scan/fai-100-controller/src/fai_100_controllerd.rs

```diff
@@ -0,0 +1,555 @@
+/// If changing from inactive to active, sends the appropriate event (keypress or system file update).
+/// If changing from active to inactive or no status change, does nothing.
+fn handle_status_response(
+    new_status: NotificationStatusResponse,
+    current_status: &mut CurrentStatus,
+    keyboard: &mut impl VirtualKeyboard,
+) -> Result<(), CommandError> {
+    let new_button = new_status.button_pressed;
+    let new_sip_status = new_status.sip_status;
+    let new_puff_status = new_status.puff_status;
+    let new_connection_status = new_status.sip_puff_device_connection_status;
+
+    // If the new button isn't currently being pressed (ie. this is a keydown event), send a keypress
+    if new_button != current_status.button_pressed {
+        // Only sends keypress when a new button is pressed.
+        // This will no-op when a button is released because new_button == ButtonSignal::NoButton
+        // get_key_for_button_signal(ButtonSignal::NoButton) returns None
+        if let Some(keypress_spec) = get_key_for_button_signal(&new_button) {
+            send_keystroke(&keypress_spec, keyboard);
+        }
+    }
+    // Update the currently pressed button so subsequent identical status reports won't trigger a second keypress
+    current_status.button_pressed = new_button;
+
+    // Write device connection status to system file so mark-scan app is aware
+    if new_connection_status != current_status.sip_puff_device_connected {
+        write_pat_connection_status(new_connection_status == SipAndPuffDeviceStatus::Connected)?;
+
+        // Explicitly reset sip and puff signals if device was just disconnected because signals are
+        // ignored once device is disconnected. See comment below.
+        let just_disconnected = new_connection_status == SipAndPuffDeviceStatus::Disconnected
+            && current_status.sip_puff_device_connected == SipAndPuffDeviceStatus::Connected;
+        if just_disconnected {
+            current_status.sip = SipAndPuffSignalStatus::Idle;
+            current_status.puff = SipAndPuffSignalStatus::Idle;
+        }
+    }
+    current_status.sip_puff_device_connected = new_connection_status;
+
+    // Only check for sip & puff actions when the device is connected. When the device is not connected
+    // the values switch meaning.
+    // ie. When a sip & puff is connected, 0x01 is sent to host when a sip
+    // is happening and 0x00 is sent when no sip is happening.
+    // When no sip & puff is connected, 0x00 is sent when no sip is happening.
+    // Therefore we must ignore all sip & puff signals unless we know the sip & puff is connected.
+    if current_status.sip_puff_device_connected == SipAndPuffDeviceStatus::Connected {
+        // Send keypress for new sip event
+        if new_sip_status == SipAndPuffSignalStatus::Active
+            && current_status.sip == SipAndPuffSignalStatus::Idle
```

Comment:
Do we know whether there's any jitter in these? Will the device handle fluctuation in the sip/puff strength and do the thresholding itself to ensure it always sends a single clean transition event? Or do we need to be concerned about transitions like Active → Active → Active → Idle → Active → Idle → Active → Idle → Idle → Idle?

Reply:
I don't actually have a sip & puff on hand at the moment (the one I had is being used for demos) so I'm not sure. I've been testing with buttons that are much more clearly binary.

The report period of the device seems to be 100ms though, so the situation you described is partially mitigated. It would depend on how fast the fluctuation happens and probably varies device to device. It's a good call out that we should test when we have access to a sip and puff device.